include/simics/dmllib.h

```diff
@@ -2256,21 +2256,21 @@ __qname(dml_qname_cache_t *cache, const char *fmt, ...)
         return (const char *)s;
 }
 
-UNUSED static const char *
-_DML_get_qname(_identity_t id, const _id_info_t *id_infos,
-               dml_qname_cache_t *cache, const char *dev_name) {
+UNUSED static char *
+_DML_get_qname_alloc(_identity_t id, const _id_info_t *id_infos,
+                     const char *dev_name) {
     _id_info_t info = id_infos[id.id - 1];
 
     const char *logname = info.logname;
 
     // In order to distinguish the device object from any other, its id_info
     // logname is always "dev", but we want its name when it comes to qname.
     if (strcmp(logname, "dev") == 0) {
-        return dev_name;
+        return strdup(dev_name);
     }
 
     if (info.dimensions == 0) {
-        return logname;
+        return strdup(logname);
```

Comment:
This one is a significant cost increase for a common shortcut: the intent was to return early from `_DML_get_qname` when things are simple; now we'll instead alloc/dealloc redundantly, copy it once more, and also consume an element of the qname ring buffer. I suggest that we instead move these shortcuts to `_DML_get_qname`, where they help.

Reply:
This would be a microoptimization. Are you really expecting it to matter?

Reply:
* That code path *is* already a micro-opt; you are preserving that micro-opt but eliminating most of the optimization from it.
* It's not only a micro-opt, it is also a defensive hack to eases the burden on the qname ring buffer hack: If you write
  ```
  sb_newf("%s %s %s %s %s", a.qname, b.qname, c.qname, d.qname, e.qname);
  ```
  then it will currently break only if *all* five objects are inside arrays. With your change it will unconditionally break.